### ISSUE TITLE
When using displayByAle, leaving insert mode resets the selected location list entry to the first entry

### ISSUE THREAD

**Issue by YorickPeterse** [neoclide/coc.nvim#2701]
## Result from CocInfo

```
## versions

vim version: NVIM v0.4.4
node version: v15.2.0
coc.nvim version: 0.0.79
coc.nvim directory: /home/yorickpeterse/Projects/general/dotfiles/dotfiles/.config/nvim/plugged/coc.nvim
term: undefined
platform: linux

## Log of coc.nvim

2020-12-10T23:00:23.433 INFO (pid:818024) [services] - registered service "jedi"
2020-12-10T23:00:23.446 INFO (pid:818024) [services] - registered service "json"
2020-12-10T23:00:23.446 INFO (pid:818024) [services] - Json language server state change: stopped => starting
2020-12-10T23:00:23.461 INFO (pid:818024) [services] - registered service "rust-analyzer"
2020-12-10T23:00:23.476 INFO (pid:818024) [language-client-index] - json started with 818040
2020-12-10T23:00:23.537 INFO (pid:818024) [services] - Json language server state change: starting => running
2020-12-10T23:00:23.541 INFO (pid:818024) [services] - service json started
2020-12-10T23:00:24.449 WARN (pid:818024) [extensions] - Extension coc-rust-analyzer activate cost more than 1s
2020-12-10T23:00:24.453 INFO (pid:818024) [plugin] - coc.nvim 0.0.79 initialized with node: v15.2.0 after 1328ms
2020-12-10T23:00:50.768 INFO (pid:818024) [willSaveHandler] - Will save cost: 61
2020-12-10T23:01:07.045 INFO (pid:818024) [willSaveHandler] - Will save cost: 84
```

## Describe the bug

I use `"diagnostic.displayByAle": true` so that errors are displayed using ALE. ALE is set up to use Vim's location list to display diagnostics. Whenever there are entries, leaving insert mode results in the selected location list entry being reset to the first one.

## Screenshots

We start off with the location list open, with entry 5 being selected:

![step1](https://user-images.githubusercontent.com/86065/101836649-1322f000-3b3e-11eb-9ef3-4f461ef0eba5.png)

Next, I switch to the buffer with the source code and enter insert mode:

![step2](https://user-images.githubusercontent.com/86065/101836680-23d36600-3b3e-11eb-9280-6d8877177add.png)

Finally I exit insert mode. Note how the location list jumps to the first entry:

![step3](https://user-images.githubusercontent.com/86065/101836721-3483dc00-3b3e-11eb-9af3-33efd3a24b08.png)

This makes it hard to jump around with commands such as `:lne`, as the current position in the location list resets every time you leave insert mode.

Note that these diagnostics are _not_ produced by coc.nvim or a language server. Instead, they are produced by an ALE linter. Disabling coc.nvim entirely prevents this bug from happening.

## Reproduce the bug

Basic vim config file:

  ```vim
let g:plug_url_format = 'git@github.com:%s.git'

call plug#begin('~/.config/nvim/plugged')

Plug 'neoclide/coc.nvim', { 'branch': 'master', 'do': 'yarn install --frozen-lockfile' }
Plug 'dense-analysis/ale'

call plug#end()

" Configure ALE
let g:ale_disable_lsp = 1
let g:ale_lint_on_enter = 0
let g:ale_fix_on_save = 1
let g:ale_fix_on_enter = 0
let g:ale_lint_on_text_changed = 'never'
let g:ale_lint_on_insert_leave = 0

" Configure coc.nvim
let g:coc_enable_locationlist = 0
  ```

And the `CocConfig` settings:

```jsonc
{
    "diagnostic.displayByAle": true
}
```

Create `/tmp/test.rb` with the following content (you'll need Ruby for this particular example):

```ruby
foo = 10
bar = 20
baz = 30
```

Start NeoVim using `nvim -u /tmp/repro.vim /tmp/test.rb`. You'll see something like this:

![step4](https://user-images.githubusercontent.com/86065/101837519-71040780-3b3f-11eb-8b5d-1206ea8a77b2.png)

Run `:ALELint`, followed by `:lopen`. You'll see something like this:

![step5](https://user-images.githubusercontent.com/86065/101837571-86793180-3b3f-11eb-95ca-58d232a7e274.png)

Enter the location list, and select an entry (e.g. the 4th one):

![step6](https://user-images.githubusercontent.com/86065/101837900-0ef7d200-3b40-11eb-8bb7-743bfce9ea63.png)

Then switch to the buffer with the Ruby code, enter insert mode, then leave insert mode. Note how the selected entry in the location list jumps back to the first one:

![step7](https://user-images.githubusercontent.com/86065/101837923-19b26700-3b40-11eb-80ac-2593a7fbfc07.png)

If you now run `:lne` it jumps to the 1st entry, not the 5th (assuming you selected the 4th one earlier).

**Comment by chemzqm** [comment]
It's expected since diagnostics is updated and ale refreshed your locationlist

**Comment by YorickPeterse** [comment]
@chemzqm Unless I'm misunderstanding, that doesn't sound quite right. The setting `g:ale_lint_on_insert_leave` is specifically disabled, and leaving insert mode doesn't trigger any new linting. As such I would expect the location list to stay as it is.

**Comment by chemzqm** [comment]
Diagnostics of coc.nvim are automatically refreshed since that's how language server, no trigger at all.
You have to not use diagnostics from coc.nvim if you want that behavior.

**Comment by chemzqm** [comment]
Or use `:CocList diagnostics` instead which doesn't refersh on diagnostics change.

**Comment by YorickPeterse** [comment]
@chemzqm The diagnostics in this case aren't produce by a language server at all. Instead, they are produced by an ALE linter. In other words, coc.nvim is not doing _any_ work here. To illustrate that this is coc.nvim and not ALE itself: if I run `autocmd! coc_nvim` to clear out all autocmds, the issue goes away (but of course parts of coc then stop working).

**Comment by chemzqm** [comment]
> coc.nvim version: 0.0.79

You're not using release branch of coc.nvim, which could be too old.

**Comment by YorickPeterse** [comment]
@chemzqm The commit I'm using from coc.nvim is `bfac5f9416e97b30a1904fa1c3b8ddb8534a29bb`, which is pretty new.

**Comment by chemzqm** [comment]
Try disable https://github.com/neoclide/coc.nvim/blob/721f227aa06ccf2f651ecc00c78502ece1d1aa6b/src/diagnostic/manager.ts#L706to see if it works.

**Comment by YorickPeterse** [comment]
@chemzqm Disabling that line stops the problem from happening. Perhaps this line should only run when there are items to display, as it currently runs even if `aleItems.length == 0`. For example, if I add `nvim.command(`echo 'ALE items: ${aleItems.length}'`)` it will print `ALE items: 0` when I exit insert mode.

**Comment by chemzqm** [comment]
I think that ALE should not refresh your location list with same diagnostics received

**Comment by chemzqm** [comment]
Or we have to prevent send same diagnostics.

**Comment by YorickPeterse** [comment]
It seems that ALE indeed jumps to the first entry. That is, `call ale#other_source#ShowResults(bufnr('.'), 'coc', [])` also triggers the problem. I'm not sure how ALE would be able to handle this though. 

Perhaps coc.nvim shouldn't call `ale#other_source#ShowResults` when a language server is not configured? That way you can still use ALE, without `ShowResults` resetting the location list. Of course if you do use a language server it will call this, but there's no way around that.

**Comment by chemzqm** [comment]
We can simply check equals before send diagnostics, which is eas,

**Comment by chemzqm** [comment]
Fixed by a6a5daf

**Comment by YorickPeterse** [comment]
@chemzqm It works, thanks! :confetti_ball: :+1:
